rust/src/lib.rs

```diff
@@ -2745,9 +2783,39 @@ pub unsafe extern "C" fn zcashlc_propose_shielding(
                         )
                     })
             })?
-            .keys()
-            .cloned()
-            .collect();
+            .into_iter();
+
+        let from_addrs = if let Some((receiver_if_none_provided, value)) = account_receivers.next()
+        {
+            // At least one transparent receiver has funds to shield.
+            if let Some(addr) = transparent_receiver {
+                if (addr == receiver_if_none_provided && value >= shielding_threshold.into())
+                    || account_receivers.any(|(a, v)| addr == a && v >= shielding_threshold.into())
+                {
+                    // The provided transparent receiver has sufficient funds to shield.
+                    [addr]
+                } else {
+                    // Insufficient funds to shield; don't create a proposal.
+                    return Ok(FfiBoxedSlice::none());
+                }
+            } else if account_receivers.next().is_none() {
+                // Only one transparent receiver has funds to shield...
+                if value >= shielding_threshold.into() {
+                    // ... and it has sufficient funds to shield.
+                    [receiver_if_none_provided]
+                } else {
+                    // Insufficient funds to shield; don't create a proposal.
+                    return Ok(FfiBoxedSlice::none());
+                }
+            } else {
+                return Err(anyhow!(
+                        "Account has more than one transparent receiver with funds to shield; this is not yet supported by the SDK. Provide a specific transparent receiver to shield funds from."
+                    ));
+            }
+        } else {
+            // There are no transparent funds to shield; don't create a proposal.
+            return Ok(FfiBoxedSlice::none());
+        };
```

Comment:
I've tried rewriting this to use the `HashMap` directly. It is almost as complex, and now has an unwrap.
```suggestion
        let from_addrs = match account_receivers.len() {
            // There are no transparent funds to shield; don't create a proposal.
            0 => return Ok(FfiBoxedSlice::none()),
            // Only one transparent receiver has funds to shield.
            1 => {
                if let Some(addr) = transparent_receiver {
                    match account_receivers.get(&addr) {
                        // The provided transparent receiver has sufficient funds to shield.
                        Some(&value) if value >= shielding_threshold.into() => [addr],
                        // No or insufficient funds to shield; don't create a proposal.
                        None | Some(_) => return Ok(FfiBoxedSlice::none()),
                    }
                } else {
                    let (addr, value) = account_receivers.into_iter().next().expect("has entry");
                    if value >= shielding_threshold.into() {
                        // The only transparent receiver has sufficient funds to shield.
                        [addr]
                    } else {
                        // Insufficient funds to shield; don't create a proposal.
                        return Ok(FfiBoxedSlice::none());
                    }
                }
            }
            _ => {
                return Err(format_err!(
                    "Account has more than one transparent receiver with funds to shield; this is not yet supported by the SDK. Provide a specific transparent receiver to shield funds from."
                ));
            }
        };
```

Reply:
What about this: 
```rust
         let from_addrs = if account_receivers.len() > 1 {
            return Err(anyhow!(
                    "Account has more than one transparent receiver with funds to shield; this is not yet supported by the SDK. Provide a specific transparent receiver to shield funds from."
                ));
        } else {
            if let Some((addr, value)) = transparent_reciver
                .and_then(|addr| account_receivers.get(addr).map(|value| (addr, value)))
                .or_else(|| account_receivers.iter().next())
            {
                if value >= shielding_threshold.into() {
                    // ... and it has sufficient funds to shield.
                    [*addr]
                } else {
                    // Insufficient funds to shield; don't create a proposal.
                    return Ok(FfiBoxedSlice::none());
                }
            } else {
                // There are no transparent funds to shield; don't create a proposal.
                return Ok(FfiBoxedSlice::none());
            }
        };
```

Reply:
I've now gone back to my iterator approach but leveraging the length-matching approach I used above for the `HashMap`, and it's much cleaner:
```suggestion
        let from_addrs = match (
            transparent_receiver,
            account_receivers.next(),
            account_receivers.next(),
        ) {
            // There are no transparent funds to shield.
            (_, None, None) => return Ok(FfiBoxedSlice::none()),
            // There is only one transparent receiver and it has insufficient funds to
            // shield. It is irrelevant whether this is the provided receiver; if it is
            // not, then the provided receiver has no funds to shield.
            (Some(_), Some((_, value)), None) if value < shielding_threshold.into() => {
                return Ok(FfiBoxedSlice::none())
            }
            // The provided transparent receiver has no funds to shield.
            (Some(addr), Some((addr_with_funds, _)), None) if addr != addr_with_funds => {
                return Ok(FfiBoxedSlice::none())
            }
            // The provided transparent receiver has sufficient funds to shield.
            (Some(addr), Some(_), None) => [addr],
            // The only transparent receiver has sufficient funds to shield.
            (None, Some((addr_with_funds, _)), None) => [addr_with_funds],
            _ => {
                return Err(format_err!(
                    "Account has more than one transparent receiver with funds to shield; this is not yet supported by the SDK. Provide a specific transparent receiver to shield funds from."
                ));
            }
        };
```

Reply:
https://github.com/Electric-Coin-Company/zcash-light-client-ffi/pull/122